### ISSUE TITLE
issue: Watch don't work with React Compiler (React 19)

### ISSUE THREAD

**Issue by Maxou44** [react-hook-form/react-hook-form#11910]
### Version Number

7.51.4

### Codesandbox/Expo snack

https://github.com/Maxou44/issue-react-compiler-react-hook-form

### Steps to reproduce

1. Clone `https://github.com/Maxou44/issue-react-compiler-react-hook-form`
2. Run `npm i --force && npm run start`
3. Open a web browser on `http://localhost:5173/`
4. Type something in the input field, the watched value is not displayed on the page

To disable the React compiler and confirm it works without the compiler, comment the line `plugins: [["babel-plugin-react-compiler", ReactCompilerConfig]],` in vite.config.mjs, relaunch `npm start` and type something in the input field

### Expected behaviour

The use of React Compiler don't allow the use of the watch() function, it should work.

### What browsers are you seeing the problem on?

Firefox, Chrome, Safari, Edge

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by Maxou44** [comment]
Since React Compiler is still experimental, I've also reported the problem on the React repository : https://github.com/facebook/react/issues/29174

**Comment by A-Shleifman** [comment]
The same applies to the formState.

```ts
export const useFieldError = () => {
  const { fieldKey } = use(SchemaFieldContext);
  const { errors } = useFormState({ name: fieldKey, exact: true });

  return get(errors, fieldKey)?.message?.toString();
};
```

This hook fails to subscribe to changes when React Compiler is enabled. I had to add `use no memo` for now.

**Comment by A-Shleifman** [comment]
`register` sometimes works, but sometimes refuses to track changes. The `register()` call has to be called with `use no memo` to work reliably.

Another hack for now:

```ts
const useRegisterProps = (name: string, props: RegisterOptions = {}) => {
  'use no memo';

  return useFormContext().register(name, props);
};
```

**Comment by bluebill1049** [comment]
```tsx
const { watch } = useForm({ defaultValues: { customerId: "", addressId: "" } });
const customerId = useMemo(() =>watch("customerId"), [watch]);
```

`watch` reference is not going to change on each update unless we change the design. almost need to remove this API just for the compiler. At this stage, consider using useWatch as a solution.

**Comment by josephsavona** [comment]
@Maxou44 thanks for reporting this here! For folks working on react-hook-form, we've written up some guidelines for how to think about designing libraries for maximum compatibility with React, including React Compiler.

As @bluebill1049 notes, using `useMemo` with `watch` will cause results to be stale. This is part of a pattern, where non-hook APIs should always be memoizable. We cover this in https://github.com/reactwg/react-compiler/discussions/10 ("Any non-hook usage of your library should be memoize-able" section). Please let us know if you have questions!

**Comment by bluebill1049** [comment]
thanks for the comment @josephsavona, is there a way we can flag the compiler to skip the memo within the library to avoid breaking change?

**Comment by Gawdfrey** [comment]
I guess ["use no memo"](https://react.dev/learn/react-compiler#use-no-memo) can be used temporarily?

**Comment by josephsavona** [comment]
The challenge is that _callers_ of the API would need to skip compilation. For your library itself you could just not compile the library - that’s your choice! But developers are going to expect that their libraries don’t block them from adopting the compiler on their own code. 

Ultimately the watch API would need to either change (so that the form object itself changes when its state changes) or be replaced w the useWatch() API.

**Comment by bluebill1049** [comment]
> For your library itself you could just not compile the library - that’s your choice! 

Could you give some more details on this? We should probably do this before introducing breaking change.

**Comment by josephsavona** [comment]
What I mean is that you can choose whether to apply React Compiler to the code that you publish. 

Regardless, though, your _users_ might wrap a call to watch() in a useMemo() (always allowed in React) or use React Compiler, which will do the equivalent of wrapping watch() calls in useMemo, and that will break their apps. 

Note that useWatch is compatible w useMemo and React Compiler, so it seems like you should be able to deprecate watch in favor of useWatch. Unless there’s something I’m missing! This is all just a heads up so that you have context on potential issues your users may face using the library.

**Comment by bluebill1049** [comment]
thanks, @josephsavona makes sense. I will seek a solution if possible without deprecating `watch`. It's a great API for DX.

**Comment by Maxou44** [comment]
If we put the watch function in a useState, and update it each time a value is changed, it should work, no ?

**Comment by bluebill1049** [comment]
> If we put the watch function in a useState, and update it each time a value is changed, it should work, no ?

`return <div>{watch('test)}</div>` should work just fine.

**Comment by mwskwong** [comment]
@bluebill1049 @josephsavona So can I conclude that, at this stage, if we still want/need to use `watch()` with React Compiler enabled, there are two workarounds:

Option 1: replace `watch()` with `useWatch()`
```diff
  const { control, watch } = useForm();
- const field = watch('field');
+ const field = useWatch({ name: 'field', control })
```

Option 2: opt out of optimization with `use no memo` for components using `watch()`
```diff
  const Component = () => {
+   'use no memo'
    const { control, watch } = useForm();
    const field = watch('field');
  }
```

Is that correct?

**Comment by nadeemc** [comment]
@mwskwong For what it's worth, only `useWatch` seems to be working consistently from me.

**Comment by josephsavona** [comment]
Yeah the most reliable option is useWatch().

**Comment by Maxou44** [comment]
What's the plan to fix this directly in React Hook Form? How can I help?

**Comment by maurer2** [comment]
Hello, should new projects with RHF prefer to use `useWatch` over of `watch` or is it too early too say?

Cheers

**Comment by josephsavona** [comment]
I would strongly recommend that you use `useWatch()` so that you avoid the problems that can occur with `watch`. Ideally the docs could be updated to reflect this guidance.

**Comment by michael-harley** [comment]
For the `useWatch` and `watch` performance, I created a simple performance checked, feel free to check here: https://codesandbox.io/p/sandbox/interesting-field-jyc2zv?file=%2Fsrc%2FApp.js

TLDR; just use `useWatch` if possible.

**Comment by jonreading81** [comment]
I have also found the same issue with `getValues` returned from `useFormContext` and replaced it this with `useWatch`.

**Comment by jonreading81** [comment]
I have also found an issue with the `errors` object from the formState retuned from `useForm`. If these is a passed as prop to a child component, the child component does not re-ender when the value changes, so. I assume the value is being stored as a ref rather than state. I have got round this by using useFormState within the child component rather than passing it in as a prop.

**Comment by titouancreach** [comment]
Do we have solution or workaround for "formState" ? 
I have the classical <SubmitButton /> that display a load indicator when this form is submitting.

```ts
export function SubmitButton({ children, ...props }: ButtonProps) {
  const context = useFormContext();

  // unsuccessful try
  const formState = useFormState({ control: context.control });

  console.log("form state", formState);
  
  // old code before the compiler that worked
  console.log("isSubmitting", context?.formState?.isSubmitting);

  return (
    <Button
      type="submit"
      {...props}
      disabled={context?.formState?.isSubmitting}
    >
      {context?.formState.isSubmitting ? (
        <Loader2 className="size-4 mr-2 animate-spin" />
      ) : null}
      {children}
    </Button>
  );
}
```
